toolchain/check/check_unit.cpp

```diff
@@ -511,9 +514,52 @@ auto CheckUnit::CheckRequiredDefinitions() -> void {
   }
 }
 
+auto CheckUnit::CheckPoisonedConcreteImplLookupQueries() -> void {
+  // Impl lookup can generate instructions (via deduce) which we don't use, as
+  // we're only generating diagnostics here, so we catch and discard them.
+  context_.inst_block_stack().Push();
+  auto poisoned_queries =
+      std::exchange(context_.poisoned_concrete_impl_lookup_queries(), {});
+  for (const auto& poison : poisoned_queries) {
+    auto witness_result =
+        EvalLookupSingleImplWitness(context_, poison.loc_id, poison.query,
+                                    poison.non_canonical_query_self_inst_id,
+                                    /*poison_concrete_results=*/false);
+    CARBON_CHECK(witness_result.has_concrete_value());
+    auto witness_id = witness_result.concrete_witness();
+    if (witness_id != poison.impl_witness) {
+      // We can get the `Impl` from the resulting witness here, which is the
+      // `Impl` that conflicts with the previous poison query.
+      auto table_id = context_.insts()
+                          .GetAs<SemIR::ImplWitness>(witness_id)
+                          .witness_table_id;
+      auto impl_id =
+          context_.insts().GetAs<SemIR::ImplWitnessTable>(table_id).impl_id;
+      const auto& impl = context_.impls().Get(impl_id);
+      auto interface_name_id =
+          context_.interfaces().Get(impl.interface.interface_id).name_id;
+      CARBON_DIAGNOSTIC(
+          PoisonedImplLookupConcreteResult, Error,
+          "found `impl` that changes the result of a previous use "
+          "of interface `{0}`",
+          SemIR::NameId);
+      auto builder =
+          emitter_.Build(impl.first_decl_id(), PoisonedImplLookupConcreteResult,
```

Comment:
Can we also note the location of the `impl` that was previously selected for the query? I think that'd be useful.

Reply:
Yes, indeed. I am struggling a bit with how to phrase this. Putting an impl in the error and a different impl in the note is a bit confusing. But making the error on the impl usage seems worse for how to explain the error. I will just.. do something, and please lmk if you see improvements.